sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1115,9 +1126,38 @@ object SQLContext {
             })
         }
     }
-    def createConverter(cls: Class[_], dataType: DataType): Any => Any = dataType match {
-      case struct: StructType => createStructConverter(cls, struct.map(_.dataType))
-      case _ => CatalystTypeConverters.createToCatalystConverter(dataType)
+    def createConverter(t: Type, dataType: DataType): Any => Any = (t, dataType) match {
+      case (cls: Class[_], struct: StructType) =>
```

Comment:
wait .. can we reuse `JavaTypeInference.serializerFor` and make a projection, rather then reimplementing whole logics here?

Reply:
https://github.com/apache/spark/blob/5264164a67df498b73facae207eda12ee133be7d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#L131-L132

We should drop the support for getter or setter only. adding @cloud-fan here as well.

Reply:
Reusing `JavaTypeInference.serializerFor` would be great, but currently it behaves a little differently. At least it doesn't support `java.lang.Iterable[_]`, so we can't use it immediately. We need to extend it to support `Iterable` (and also `deserializerFor`).

Reply:
Hm, about we fix them together while we are here?

I also checked another difference which is beans without getter and/or setter but I think this is something we should fix in 3.0.

Reply:
Frankly, I was not really sure about serializing sets as arrays as the result stops behaving like a set, but I found a PR (#18416) where this seems to have been permitted, so I will go ahead and add that.